lib/internal/loader/Loader.js

```diff
@@ -23,22 +24,48 @@ function getBase() {
   }
 }
 
+/* A Loader instance is used as the main entry point for loading ES modules.
+ * Currently, this is a singleton -- there is only one used for loading
+ * the main module and everything in its dependency graph. */
 class Loader {
-  constructor(base = getBase()) {
-    this.moduleMap = new ModuleMap();
+  constructor(base = getURLStringForCwd()) {
     if (typeof base !== 'string') {
       throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'base', 'string');
     }
+
+    this.moduleMap = new ModuleMap();
     this.base = base;
-    this.resolver = ModuleRequest.resolve.bind(null);
+    // The resolver has the signature
+    //   (specifier : string, parentURL : string, defaultResolve)
+    //       -> Promise<{ url : string,
+    //                    format: anything in Loader.validFormats }>
+    // where defaultResolve is ModuleRequest.resolve (having the same
+    // signature itself).
+    // If `.format` on the returned value is 'dynamic', .dynamicInstantiate
+    // will be used as described below.
+    this.resolver = ModuleRequest.resolve;
+    // This hook is only called when resolve(...).format is 'dynamic' and has
+    // the signature
+    //   (url : string) -> Promise<{ exports: { ... }, execute: function }>
+    // Where `exports` is an object whose property names define the exported
+    // names of the generated module. `execute` is a function that receives
+    // an object with the same keys as `exports`, whose values are get/set
+    // functions for the actual exported values.
     this.dynamicInstantiate = undefined;
   }
 
+  static get validFormats() {
+    return ['esm', 'cjs', 'builtin', 'addon', 'json', 'dynamic'];
```

Comment:
This could be a static property instead of a getter, to avoid recreating such an array every time.

Reply:
yup, makes sense!